A47-xds-federation.md

```diff
@@ -0,0 +1,682 @@
+  we would have used prior to this design.
+
+### Implementation Changes
+
+This section describes the changes needed to implement this design.
+This information is relevant to those implementing this functionality in
+gRPC, but it is not useful to those simply wishing to know how to use
+the functionality in an application.
+
+#### `xds` Resolver Changes
+
+The xds resolver will be responsible for using the bootstrap information
+above to determine what Listener resource name to subscribe to for a
+given target URI.  The logic will be (pseudocode):
+
+```
+target_hostname = StripLeadingSlash(target_uri.path)
+if target_uri.authority is set:
+  authority_config = bootstrap.LookupAuthority(target_uri.authority)
+  if authority_config not found:
+    handle as invalid target URI
+  else:
+    template = authority_config.client_listener_resource_name_template
+    if template not set:
+      template = ("xdstp://" + target_uri.authority +
+                  "/envoy.config.listener.v3.Listener/%s")
+    xds_resource_name = ExpandPercentS(template,
+                                       PercentEncode(target_hostname))
+else:  # target_uri.authority not set
+  template = bootstrap.client_default_listener_resource_name_template
+  if template not set:
+    template = "%s"
+  if template starts with "xdstp:":
+    target_hostname = PercentEncode(target_hostname)
+  xds_resource_name = ExpandPercentS(template, target_hostname)
+```
+
+#### xDS-Enabled gRPC Server Changes
+
+Similarly to the xds resolver on the client side, the xDS-enabled gRPC
+server will need to be changed such that if the
+`server_listener_resource_name_template` field in the bootstrap file
+starts with `xdstp:`, then it will percent-encode the listening address
+before replacing the `%s` token in the template.
+
+#### `XdsClient` Changes
+
+This design will require a number of changes in `XdsClient`.
+
+##### Server Channel Map
+
+Currently, the `XdsClient` contains a single server channel.  This will be
+replaced with a map of server channels, keyed by the server definition
+from the bootstrap file (i.e., target URI, channel creds, and all known
+server features -- unknown features can be removed).  This key is used
+to de-dup servers, so that if multiple authorities use the same server,
+we do not create duplicate channels.
+
+We don't want to proactively connect a channel that is not actually
+being used.  Implementations can either proactively create the channel
+up-front and depend on idleness behavior, or they can lazily create the
+channel when it's actually needed.  Similarly, if a channel stops being
+used, implementations can either depend on idleness behavior to tear
+down the connections, or they can ref-count the channel and shut it down
+when it's no longer needed.  (The expected tear-down after being idle is
+O(5-15m).)
+
+##### Resource Cache
+
+The resource cache in the `XdsClient` is currently a two-level map, keyed
```

Comment:
What's the reason for the extra level? With the new style resource names, shouldn't every resource have a unique name?
We can keep the level for resource type, but mainly for the convenience, it won't be necessary anymore.

Reply:
Two reasons:
1. Memory efficiency.  It seems wasteful to replicate the same authority string many times in the cache.  (Note that this same argument applies to resource type.)
2. We need a place to store some per-authority data anyway, such as the server (or, eventually, list of servers) to use for that authority, and a ref to the server channel that we're currently using for it.

Also, note that the resource type level is still needed for old-style names, because a control plane could still use the same old-style resource name for multiple resource types.

Reply:
So this is more for efficiency, not correctness, is that right?

I was thinking about making all the authorities share one cache, so all the resources are kept in one place (in one two-level map, not with the extra level of authority).
It simplifies CSDS a bit.
And one goal of the new-style resource name is to make it easier for caching, so I thought this should be a reasonable choice.

One complexity I got is for the SotW APIs (LDS and CDS), those needs per authority tracking to know which resource is removed. But for RDS/EDS, the implementation before federation should just work.

Reply:
We still need a place to store the list of servers (and their associated configuration) to use for each authority, and we need a place to store a ref to the current channel being used for each authority.  If you don't store that information here, where will you store it?  Just have another map for it?

If you have another way of dealing with the above, then yes, it's "just" for efficiency, not correctness.  However, given the price of memory these days, I think the efficiency argument is actually a significant one.  If you just store the literal resource names as strings, then you're going to be duplicating the authority and resource type strings in memory many, many times.

> It simplifies CSDS a bit.

For us, using a multi-level map didn't make a big difference for CSDS.  We basically needed two nested loops instead of just one, but it wasn't really that much more complicated.

> And one goal of the new-style resource name is to make it easier for caching, so I thought this should be a reasonable choice.

To be clear, the goal of the new-style naming is actually just to provide a global naming scheme so that clients can get different resources from different servers without having any naming conflicts.  It's not really intended to make implementations simpler; in fact, as evidenced by this discussion, implementations will likely get a bit more complex. :)

> One complexity I got is for the SotW APIs (LDS and CDS), those needs per-server tracking to know which resource is removed.

Right.  I think that's another argument for having a two-level map: it avoids the need to iterate over every resource for every LDS or CDS response.  If you have a large number of authorities, you should ideally be able to look only at the resources for the authorities that are currently using the server that you're talking to.

In C-core, we are currently iterating through the list of authorities and skipping the ones whose server channel is not the channel that we received the response on.  And if we wind up in a situation where there are a very large number of authorities, we could optimize that further by keeping a list of authorities to look at in the object that holds the server channel.